src/core/remio.c

```diff
@@ -0,0 +1,668 @@
+                if (device == NULL) {
+                    ERROR("Failed allocating Remote I/O device node");
+                }
+                device->id = dev->id;
+                list_init(&device->request_event_list);
+                list_push(&remio_device_list, (node_t*)device);
+                backend_cnt++;
+                devices[dev->id][REMIO_DEV_BACKEND] = (int)vm_id;
+            } else {
+                frontend_cnt++;
+                devices[dev->id][REMIO_DEV_FRONTEND] = (int)vm_id;
+            }
+        }
+    }
+
+    /** Check if there is a 1-to-1 mapping between a Remote I/O backend and Remote I/O frontend */
+    if (backend_cnt != frontend_cnt) {
+        ERROR("There is no 1-to-1 mapping between a Remote I/O backend and Remote I/O frontend");
+    }
+
+    /** Update the Remote I/O device configuration */
+    for (size_t vm_id = 0; vm_id < config.vmlist_size; vm_id++) {
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+            struct remio_dev* dev = &vm_config->platform.remio_devs[i];
+            list_foreach (remio_device_list, struct remio_device, device) {
+                if (dev->id == device->id) {
+                    if (dev->type == REMIO_DEV_BACKEND) {
+                        device->config.backend.vm_id = vm_id;
+                        device->config.backend.interrupt = dev->interrupt;
+                        device->config.backend.cpu_id = (cpuid_t)-1;
+                    } else {
+                        device->config.frontend.vm_id = vm_id;
+                        device->config.frontend.interrupt = dev->interrupt;
+                        device->config.frontend.cpu_id = (cpuid_t)-1;
+                    }
+                }
+            }
+        }
+    }
+
+    /** Initialize the Remote I/O requests array */
+    for (size_t cpu_idx = 0; cpu_idx < REMIO_CPU_NUM; cpu_idx++) {
+        for (size_t vcpu_idx = 0; vcpu_idx < REMIO_VCPU_NUM; vcpu_idx++) {
+            remio_set_request_state(cpu_idx, vcpu_idx, REMIO_STATE_FREE);
+        }
+    }
+}
+
+void remio_assign_vm_cpus(struct vm* vm)
+{
+    list_foreach (remio_device_list, struct remio_device, io_device) {
+        if (vm->id == io_device->config.backend.vm_id) {
+            io_device->config.backend.cpu_id = min(io_device->config.backend.cpu_id, cpu()->id);
+        } else if (vm->id == io_device->config.frontend.vm_id) {
+            io_device->config.frontend.cpu_id = min(io_device->config.frontend.cpu_id, cpu()->id);
+        }
+    }
+}
+
+/**
+ * @brief Finds the Remote I/O device associated with a VM based on the Remote I/O device ID
+ * @param vm Pointer to the VM structure
+ * @param id Remote I/O device ID
+ * @return Returns the Remote I/O device or NULL if the device was not found
+ */
+static struct remio_device* remio_find_vm_io_dev_by_id(struct vm* vm, unsigned long id)
```

Comment:
I suggest this name, but now realize it should be `remio_find_vm_io_dev_by_id`
other "_io_dev" instances should be changed similarly

Reply:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/29c3e19557de3c3aaa1cac56e49b8b535d53fe0c